File: ios/Sources/App/Services/CloudVaultService.swift
import CryptoKit
import Foundation
import SwiftData

enum CloudVaultError: Error, LocalizedError {
    case invalidResponse
    case server(String)
    case emptyVault

    var errorDescription: String? {
        switch self {
        case .invalidResponse: return "The cloud vault returned an invalid response."
        case .server(let message): return message
        case .emptyVault: return "No encrypted vault has been uploaded yet."
        }
    }
}

private struct HostVaultRecord: Codable {
    var id: UUID; var label: String; var address: String; var port: Int; var username: String
    var authMethod: HostAuthMethod; var identityID: UUID?; var startupSnippet: String?
    var groupName: String?; var tags: [String]; var themeName: String?; var password: String?
}

private struct IdentityVaultRecord: Codable {
    var id: UUID; var label: String; var keyType: IdentityKeyType; var fingerprint: String
    var publicKey: String; var privateKey: String?; var passphrase: String?
}

private struct SnippetVaultRecord: Codable {
    var id: UUID; var name: String; var command: String; var runOnConnect: Bool
}

private struct VaultSnapshot: Codable {
    var version = 1
    var createdAt = Date()
    var hosts: [HostVaultRecord]
    var identities: [IdentityVaultRecord]
    var snippets: [SnippetVaultRecord]
}

private struct AuthResponse: Decodable { let token: String }
private struct VaultEnvelope: Codable { let salt: String; let nonce: String; let ciphertext: String }
private struct VaultReadResponse: Decodable { let revision: Int; let vault: VaultEnvelope? }
private struct VaultWriteRequest: Encodable { let revision: Int; let vault: VaultEnvelope }
private struct VaultWriteResponse: Decodable { let revision: Int }

@MainActor
final class CloudVaultService {
    static let shared = CloudVaultService()
    private let baseURL = URL(string: "https://masystem.co.in/termvault-api")!

    func authenticate(email: String, password: String, register: Bool) async throws {
        var request = URLRequest(url: baseURL.appending(path: register ? "v1/register" : "v1/login"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(["email": email, "password": password])
        let response: AuthResponse = try await perform(request)
        try KeychainService.setCloudToken(response.token)
    }

    func upload(hosts: [Host], identities: [Identity], snippets: [Snippet], password: String) async throws -> Int {
        let snapshot = VaultSnapshot(
            hosts: hosts.map { host in
                HostVaultRecord(
                    id: host.id, label: host.label, address: host.address, port: host.port,
                    username: host.username, authMethod: host.authMethod, identityID: host.identityID,
                    startupSnippet: host.startupSnippet, groupName: host.groupName, tags: host.tags,
                    themeName: host.themeName, password: try? KeychainService.password(for: host)
                )
            },
            identities: identities.map { identity in
                IdentityVaultRecord(
                    id: identity.id, label: identity.label, keyType: identity.keyType,
                    fingerprint: identity.fingerprint, publicKey: identity.publicKey,
                    privateKey: try? KeychainService.privateKeyPEM(for: identity),
                    passphrase: try? KeychainService.passphrase(for: identity)
                )
            },
            snippets: snippets.map { SnippetVaultRecord(id: $0.id, name: $0.name, command: $0.command, runOnConnect: $0.runOnConnect) }
        )
        let current = try await readEnvelope()
        let salt = Data((0..<16).map { _ in UInt8.random(in: .min ... .max) })
        let key = deriveKey(password: password, salt: salt)
        let plaintext = try JSONEncoder.vaultEncoder.encode(snapshot)
        let sealed = try AES.GCM.seal(plaintext, using: key)
        let envelope = VaultEnvelope(
            salt: salt.base64EncodedString(), nonce: Data(sealed.nonce).base64EncodedString(),
            ciphertext: sealed.ciphertext.base64EncodedString() + "." + sealed.tag.base64EncodedString()
        )
        var request = try authorizedRequest(path: "v1/vault", method: "PUT")
        request.httpBody = try JSONEncoder().encode(VaultWriteRequest(revision: current.revision, vault: envelope))
        let result: VaultWriteResponse = try await perform(request)
        return result.revision
    }

    func restore(into context: ModelContext, password: String) async throws -> Int {
        let result = try await readEnvelope()
        guard let envelope = result.vault else { throw CloudVaultError.emptyVault }
        let snapshot = try decrypt(envelope, password: password)
        let existingHosts = try context.fetch(FetchDescriptor<Host>())
        let existingIdentities = try context.fetch(FetchDescriptor<Identity>())
        let existingSnippets = try context.fetch(FetchDescriptor<Snippet>())

        for record in snapshot.identities {
            let value = existingIdentities.first { $0.id == record.id } ?? Identity(
                id: record.id, label: record.label, keyType: record.keyType,
                fingerprint: record.fingerprint, publicKey: record.publicKey,
                hasPassphrase: record.passphrase != nil
            )
            value.label = record.label; value.keyType = record.keyType
            value.fingerprint = record.fingerprint; value.publicKey = record.publicKey
            value.hasPassphrase = record.passphrase != nil
            if !existingIdentities.contains(where: { $0.id == record.id }) { context.insert(value) }
            if let privateKey = record.privateKey { try KeychainService.setPrivateKeyPEM(privateKey, for: value) }
            if let passphrase = record.passphrase { try KeychainService.setPassphrase(passphrase, for: value) }
        }
        for record in snapshot.hosts {
            let value = existingHosts.first { $0.id == record.id } ?? Host(
                id: record.id, label: record.label, address: record.address, port: record.port,
                username: record.username, authMethod: record.authMethod, identityID: record.identityID
            )
            value.label = record.label; value.address = record.address; value.port = record.port
            value.username = record.username; value.authMethod = record.authMethod
            value.identityID = record.identityID; value.startupSnippet = record.startupSnippet
            value.groupName = record.groupName; value.tags = record.tags; value.themeName = record.themeName
            if !existingHosts.contains(where: { $0.id == record.id }) { context.insert(value) }
            if let password = record.password { try KeychainService.setPassword(password, for: value) }
        }
        for record in snapshot.snippets {
            let value = existingSnippets.first { $0.id == record.id } ?? Snippet(
                id: record.id, name: record.name, command: record.command, runOnConnect: record.runOnConnect
            )
            value.name = record.name; value.command = record.command; value.runOnConnect = record.runOnConnect
            if !existingSnippets.contains(where: { $0.id == record.id }) { context.insert(value) }
        }
        try context.save()
        return result.revision
    }

    private func readEnvelope() async throws -> VaultReadResponse {
        try await perform(authorizedRequest(path: "v1/vault", method: "GET"))
    }

    private func decrypt(_ envelope: VaultEnvelope, password: String) throws -> VaultSnapshot {
        guard let salt = Data(base64Encoded: envelope.salt),
              let nonceData = Data(base64Encoded: envelope.nonce) else { throw CloudVaultError.invalidResponse }
        let parts = envelope.ciphertext.split(separator: ".")
        guard parts.count == 2, let ciphertext = Data(base64Encoded: String(parts[0])),
              let tag = Data(base64Encoded: String(parts[1])) else { throw CloudVaultError.invalidResponse }
        let box = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: nonceData), ciphertext: ciphertext, tag: tag)
        let plaintext = try AES.GCM.open(box, using: deriveKey(password: password, salt: salt))
        return try JSONDecoder.vaultDecoder.decode(VaultSnapshot.self, from: plaintext)
    }

    private func deriveKey(password: String, salt: Data) -> SymmetricKey {
        HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: Data(password.utf8)), salt: salt,
            info: Data("TermVault encrypted cloud vault v1".utf8), outputByteCount: 32
        )
    }

    private func authorizedRequest(path: String, method: String) throws -> URLRequest {
        var request = URLRequest(url: baseURL.appending(path: path))
        request.httpMethod = method
        request.setValue("Bearer \(try KeychainService.cloudToken())", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return request
    }

    private func perform<Value: Decodable>(_ request: URLRequest) async throws -> Value {
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw CloudVaultError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            let value = (try? JSONSerialization.jsonObject(with: data) as? [String: String])?["error"]
            throw CloudVaultError.server(value ?? "Cloud vault request failed (HTTP \(http.statusCode)).")
        }
        return try JSONDecoder().decode(Value.self, from: data)
    }
}

private extension JSONEncoder {
    static var vaultEncoder: JSONEncoder { let value = JSONEncoder(); value.dateEncodingStrategy = .iso8601; return value }
}

private extension JSONDecoder {
    static var vaultDecoder: JSONDecoder { let value = JSONDecoder(); value.dateDecodingStrategy = .iso8601; return value }
}

File: android/app/src/main/kotlin/dev/termvault/app/cloud/CloudVaultService.kt
package dev.termvault.app.cloud

import dev.termvault.app.data.db.HostDao
import dev.termvault.app.data.db.HostEntity
import dev.termvault.app.data.db.IdentityDao
import dev.termvault.app.data.db.IdentityEntity
import dev.termvault.app.data.db.SnippetDao
import dev.termvault.app.data.db.SnippetEntity
import dev.termvault.app.security.SecretStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.HKDFParameters
import retrofit2.HttpException
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import java.security.SecureRandom
import java.time.Instant
import java.util.UUID
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

private fun b64Encode(bytes: ByteArray): String = android.util.Base64.encodeToString(bytes, android.util.Base64.NO_WRAP)
private fun b64Decode(text: String): ByteArray = android.util.Base64.decode(text, android.util.Base64.NO_WRAP)

class CloudVaultError(message: String) : Exception(message)
class EmptyVaultException : Exception("No encrypted vault has been uploaded yet.")

/**
 * Encrypted cross-device sync of hosts/identities/snippets against the real
 * `backend/termvault/server.py` API — mirrors iOS `CloudVaultService.swift`
 * exactly, including the AES-256-GCM + HKDF-SHA256 envelope format, so
 * vaults uploaded from one platform can be restored on the other.
 */
class CloudVaultService(
    private val secretStore: SecretStore,
    private val hostDao: HostDao,
    private val identityDao: IdentityDao,
    private val snippetDao: SnippetDao,
) {
    private val json = Json { ignoreUnknownKeys = true }

    private val api: CloudVaultApi = Retrofit.Builder()
        .baseUrl("https://masystem.co.in/termvault-api/")
        .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
        .build()
        .create(CloudVaultApi::class.java)

    suspend fun authenticate(email: String, password: String, register: Boolean) = withContext(Dispatchers.IO) {
        val response = perform { if (register) api.register(AuthRequest(email, password)) else api.login(AuthRequest(email, password)) }
        secretStore.setCloudToken(response.token)
    }

    suspend fun upload(password: String): Int = withContext(Dispatchers.IO) {
        val hosts = hostDao.getAll()
        val identities = identityDao.getAll()
        val snippets = snippetDao.getAll()

        val snapshot = VaultSnapshot(
            createdAt = Instant.now().toString(),
            hosts = hosts.map { host ->
                HostVaultRecord(
                    id = host.id.toString(),
                    label = host.label,
                    address = host.address,
                    port = host.port,
                    username = host.username,
                    authMethod = host.authMethodRaw,
                    identityID = host.identityId?.toString(),
                    jumpHostID = host.jumpHostId?.toString(),
                    startupSnippet = host.startupSnippet,
                    groupName = host.groupName,
                    tags = host.tags,
                    themeName = host.themeName,
                    password = secretStore.password(host),
                )
            },
            identities = identities.map { identity ->
                IdentityVaultRecord(
                    id = identity.id.toString(),
                    label = identity.label,
                    keyType = identity.keyTypeRaw,
                    fingerprint = identity.fingerprint,
                    publicKey = identity.publicKey,
                    privateKey = secretStore.privateKeyPem(identity),
                    passphrase = secretStore.passphrase(identity),
                )
            },
            snippets = snippets.map { SnippetVaultRecord(it.id.toString(), it.name, it.command, it.runOnConnect) },
        )

        val current = readEnvelope()
        val salt = ByteArray(16).also { SecureRandom().nextBytes(it) }
        val key = deriveKey(password, salt)
        val plaintext = json.encodeToString(VaultSnapshot.serializer(), snapshot).toByteArray(Charsets.UTF_8)
        val (ciphertext, tag, nonce) = encrypt(plaintext, key)

        val envelope = VaultEnvelope(
            salt = b64Encode(salt),
            nonce = b64Encode(nonce),
            ciphertext = b64Encode(ciphertext) + "." + b64Encode(tag),
        )
        val result = perform { api.putVault("Bearer ${cloudToken()}", VaultWriteRequest(current.revision, envelope)) }
        result.revision
    }

    /** Decrypts the stored vault and applies it into Room + [SecretStore], mirroring iOS's SwiftData merge. */
    suspend fun restore(password: String): Int = withContext(Dispatchers.IO) {
        val result = readEnvelope()
        val envelope = result.vault ?: throw EmptyVaultException()
        val snapshot = decrypt(envelope, password)

        for (record in snapshot.identities) {
            val id = UUID.fromString(record.id)
            val value = IdentityEntity(
                id = id,
                label = record.label,
                keyTypeRaw = record.keyType,
                fingerprint = record.fingerprint,
                publicKey = record.publicKey,
                hasPassphrase = record.passphrase != null,
            )
            identityDao.upsert(value)
            record.privateKey?.let { secretStore.setPrivateKeyPem(it, value) }
            record.passphrase?.let { secretStore.setPassphrase(it, value) }
        }
        for (record in snapshot.hosts) {
            val id = UUID.fromString(record.id)
            val value = HostEntity(
                id = id,
                label = record.label,
                address = record.address,
                port = record.port,
                username = record.username,
                authMethodRaw = record.authMethod,
                identityId = record.identityID?.let { UUID.fromString(it) },
                jumpHostId = record.jumpHostID?.let { UUID.fromString(it) },
                startupSnippet = record.startupSnippet,
                groupName = record.groupName,
                tags = record.tags,
                themeName = record.themeName,
            )
            hostDao.upsert(value)
            record.password?.let { secretStore.setPassword(it, value) }
        }
        for (record in snapshot.snippets) {
            snippetDao.upsert(
                SnippetEntity(
                    id = UUID.fromString(record.id),
                    name = record.name,
                    command = record.command,
                    runOnConnect = record.runOnConnect,
                )
            )
        }
        result.revision
    }

    private suspend fun readEnvelope(): VaultReadResponse = perform { api.getVault("Bearer ${cloudToken()}") }

    private fun cloudToken(): String = secretStore.cloudToken() ?: throw CloudVaultError("Not signed in to Cloud Vault.")

    private fun decrypt(envelope: VaultEnvelope, password: String): VaultSnapshot {
        val salt = b64Decode(envelope.salt)
        val nonce = b64Decode(envelope.nonce)
        val parts = envelope.ciphertext.split(".")
        if (parts.size != 2) throw CloudVaultError("The cloud vault returned an invalid response.")
        val ciphertext = b64Decode(parts[0])
        val tag = b64Decode(parts[1])
        val key = deriveKey(password, salt)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        val plaintext = cipher.doFinal(ciphertext + tag)
        return json.decodeFromString(VaultSnapshot.serializer(), String(plaintext, Charsets.UTF_8))
    }

    /** Returns (ciphertext, tag, nonce) — Java's GCM cipher appends the 16-byte tag to the ciphertext; this splits them apart to match iOS's separate `ciphertext`/`tag` wire fields. */
    private fun encrypt(plaintext: ByteArray, key: ByteArray): Triple<ByteArray, ByteArray, ByteArray> {
        val nonce = ByteArray(12).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        val output = cipher.doFinal(plaintext)
        val ciphertext = output.copyOfRange(0, output.size - 16)
        val tag = output.copyOfRange(output.size - 16, output.size)
        return Triple(ciphertext, tag, nonce)
    }

    /** HKDF-SHA256, matching CryptoKit's `HKDF<SHA256>.deriveKey` used on iOS: extract with [salt], expand with the fixed info string, 32-byte output. */
    private fun deriveKey(password: String, salt: ByteArray): ByteArray {
        val generator = HKDFBytesGenerator(SHA256Digest())
        generator.init(HKDFParameters(password.toByteArray(Charsets.UTF_8), salt, "TermVault encrypted cloud vault v1".toByteArray(Charsets.UTF_8)))
        val output = ByteArray(32)
        generator.generateBytes(output, 0, 32)
        return output
    }

    private suspend fun <T> perform(call: suspend () -> T): T {
        try {
            return call()
        } catch (e: HttpException) {
            val body = e.response()?.errorBody()?.string()
            val message = body?.let { runCatching { json.decodeFromString(ErrorResponse.serializer(), it).error }.getOrNull() }
            throw CloudVaultError(message ?: "Cloud vault request failed (HTTP ${e.code()}).")
        }
    }
}
